src/libbson/CMakeLists.txt

```diff
@@ -119,6 +119,38 @@ else ()
    set (BSON_BYTE_ORDER 1234)
 endif ()
 
+# 1. Normally, `defined(_XOPEN_SOURCE) && _XOPEN_SOURCE >= 600` should be enough
+#    to detect if an XSI-compliant `strerror_r` is available.
+# 2. However, GNU provides an incompatible `strerror_r` as an extension,
+#    requiring an additional test for `!defined(_GNU_SOURCE)`.
+# 3. However, musl does not support the GNU extension even when `_GNU_SOURCE` is
+#    defined, preventing `!defined(_GNU_SOURCE)` from being a portable
+#    XSI-compliance test.
+# 4. However, musl does not provide a feature test macro to detect compilation
+#    with musl, and workarounds that use internal glibc feature test macros such
+#    as `defined(__USE_GNU)` are explicitly forbidden by glibc documentation.
+# 5. Therefore, give up on using feature test macros: use `CheckCSourceCompiles`
+#    to test if the current `strerror_r` is XSI-compliant if present.
+if (NOT WIN32)
+   CHECK_C_SOURCE_COMPILES(
+      [[
+         #include <string.h>
+         int test(int errnum, char *buf, size_t buflen) {
+            return strerror_r (errnum, buf, buflen);
+         }
+         int main(void) {}
+      ]]
+      BSON_HAVE_XSI_STRERROR_R
+      FAIL_REGEX "int-conversion"
+   )
+endif ()
+
+if (BSON_HAVE_XSI_STRERROR_R)
+   set(BSON_HAVE_XSI_STRERROR_R 1)
+else()
+   set(BSON_HAVE_XSI_STRERROR_R 0)
+endif ()
```

Comment:
> gave up on using feature test macros and instead opted to use CMake's CheckCSourceCompiles module to obtain the definitive truth.

This does seem like the most reliable solution. Since this solution won't apply to the PHP driver, I created [PHPC-2263](https://jira.mongodb.org/browse/PHPC-2263) to port this to Autotools.